File: Navigation/nav2_option/g1_sdk_bridge/tests/test_ipc_transport.py
import os
import socket
import tempfile
import time
import unittest

import _pathfix  # noqa: F401
from ipc_transport import PeerClosed, SeqPacketServer, connect_client


def _sock_path(tmpdir: str, name: str) -> str:
    # AF_UNIXのパス長制限(108バイト前後)に収まるよう短い名前にする
    return os.path.join(tmpdir, name)


class TestSeqPacketRoundtrip(unittest.TestCase):
    def setUp(self):
        self._tmpdir = tempfile.mkdtemp(prefix="g1ipc_")
        self._path = _sock_path(self._tmpdir, "cmd.sock")
        self._server = SeqPacketServer(self._path, max_payload=64)

    def tearDown(self):
        self._server.close()

    def test_client_to_server_roundtrip(self):
        client = connect_client(self._path, max_payload=64, timeout=1.0)
        server_ep = self._server.accept_blocking(timeout=1.0)

        client.send_latest(b"hello")
        # ノンブロッキング受信なので、届くまで少し待つ
        deadline = time.monotonic() + 1.0
        received = None
        while time.monotonic() < deadline:
            received = server_ep.recv_latest()
            if received is not None:
                break
            time.sleep(0.005)
        self.assertEqual(received, b"hello")

        client.close()
        server_ep.close()

    def test_recv_latest_drains_and_returns_only_newest(self):
        client = connect_client(self._path, max_payload=64, timeout=1.0)
        server_ep = self._server.accept_blocking(timeout=1.0)

        for i in range(5):
            ok = client.send_latest(f"msg{i}".encode())
            self.assertTrue(ok)
        time.sleep(0.05)  # 全メッセージがカーネルバッファに届くのを待つ

        # 5件送っても、recv_latestは「最新の1件」だけを返す(latest-value semantics)
        received = server_ep.recv_latest()
        self.assertEqual(received, b"msg4")

        client.close()
        server_ep.close()

    def test_peer_close_detected_on_recv(self):
        client = connect_client(self._path, max_payload=64, timeout=1.0)
        server_ep = self._server.accept_blocking(timeout=1.0)

        client.close()
        time.sleep(0.05)
        with self.assertRaises(PeerClosed):
            # recv_latestは即座に例外を出さないことがあるためリトライする
            deadline = time.monotonic() + 1.0
            while time.monotonic() < deadline:
                server_ep.recv_latest()
                time.sleep(0.01)

        server_ep.close()

    def test_peer_close_detected_on_send(self):
        client = connect_client(self._path, max_payload=64, timeout=1.0)
        server_ep = self._server.accept_blocking(timeout=1.0)

        server_ep.close()
        time.sleep(0.05)
        with self.assertRaises(PeerClosed):
            deadline = time.monotonic() + 1.0
            while time.monotonic() < deadline:
                client.send_latest(b"ping")
                time.sleep(0.01)

        client.close()

    def test_send_drops_silently_when_buffer_full(self):
        client = connect_client(self._path, max_payload=64, timeout=1.0)
        server_ep = self._server.accept_blocking(timeout=1.0)
        # 送信バッファを極端に小さくして、受信側が読まない状態ですぐ満杯にする
        client._sock.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 1)

        results = [client.send_latest(f"m{i}".encode()) for i in range(50)]
        # 全部は入りきらず、どこかでFalse(破棄)が発生するはず
        self.assertIn(False, results, "送信バッファ満杯時にFalseで破棄される設計だが、一度もFalseにならなかった")

        client.close()
        server_ep.close()


if __name__ == "__main__":
    unittest.main()

File: Navigation/nav2_stable/g1_sdk_bridge_cpp/tests/test_ipc_transport.cpp
#include "g1_sdk_bridge/ipc_transport.hpp"

#include <fcntl.h>
#include <gtest/gtest.h>
#include <sys/socket.h>
#include <unistd.h>

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <memory>
#include <stdexcept>
#include <string>
#include <thread>

using namespace g1_sdk_bridge;

namespace {

class TempSockPath {
public:
    TempSockPath() {
        char tmpl[] = "/tmp/g1ipc_test_XXXXXX";
        const char* dir = mkdtemp(tmpl);
        if (dir == nullptr) {
            throw std::runtime_error("mkdtemp失敗");
        }
        dir_ = dir;
        path_ = dir_ + "/cmd.sock";
    }
    ~TempSockPath() {
        std::remove(path_.c_str());
        ::rmdir(dir_.c_str());
    }
    const std::string& path() const { return path_; }

private:
    std::string dir_;
    std::string path_;
};

void SleepMs(int ms) { std::this_thread::sleep_for(std::chrono::milliseconds(ms)); }

}  // namespace

class SeqPacketRoundtripTest : public ::testing::Test {
protected:
    void SetUp() override { server_ = std::make_unique<SeqPacketServer>(tmp_.path(), 64); }

    TempSockPath tmp_;
    std::unique_ptr<SeqPacketServer> server_;
};

TEST_F(SeqPacketRoundtripTest, ClientToServerRoundtrip) {
    SeqPacketEndpoint client = ConnectClient(tmp_.path(), 64, 1.0);
    SeqPacketEndpoint server_ep = server_->AcceptBlocking(1.0);

    const std::string msg = "hello";
    ASSERT_TRUE(client.SendLatest(msg.data(), msg.size()));

    std::optional<std::vector<std::uint8_t>> received;
    const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(1);
    while (std::chrono::steady_clock::now() < deadline) {
        received = server_ep.RecvLatest();
        if (received.has_value()) break;
        SleepMs(5);
    }
    ASSERT_TRUE(received.has_value());
    EXPECT_EQ(std::string(received->begin(), received->end()), msg);
}

TEST_F(SeqPacketRoundtripTest, RecvLatestDrainsAndReturnsOnlyNewest) {
    SeqPacketEndpoint client = ConnectClient(tmp_.path(), 64, 1.0);
    SeqPacketEndpoint server_ep = server_->AcceptBlocking(1.0);

    for (int i = 0; i < 5; ++i) {
        const std::string msg = "msg" + std::to_string(i);
        ASSERT_TRUE(client.SendLatest(msg.data(), msg.size()));
    }
    SleepMs(50);  // 全メッセージがカーネルバッファに届くのを待つ

    // 5件送っても、RecvLatestは「最新の1件」だけを返す(latest-value semantics)
    const auto received = server_ep.RecvLatest();
    ASSERT_TRUE(received.has_value());
    EXPECT_EQ(std::string(received->begin(), received->end()), "msg4");
}

TEST_F(SeqPacketRoundtripTest, PeerCloseDetectedOnRecv) {
    SeqPacketEndpoint client = ConnectClient(tmp_.path(), 64, 1.0);
    SeqPacketEndpoint server_ep = server_->AcceptBlocking(1.0);

    client.Close();
    SleepMs(50);

    bool threw = false;
    const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(1);
    while (std::chrono::steady_clock::now() < deadline) {
        try {
            server_ep.RecvLatest();
        } catch (const PeerClosed&) {
            threw = true;
            break;
        }
        SleepMs(10);
    }
    EXPECT_TRUE(threw);
}

TEST_F(SeqPacketRoundtripTest, PeerCloseDetectedOnSend) {
    SeqPacketEndpoint client = ConnectClient(tmp_.path(), 64, 1.0);
    SeqPacketEndpoint server_ep = server_->AcceptBlocking(1.0);

    server_ep.Close();
    SleepMs(50);

    bool threw = false;
    const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(1);
    const std::string msg = "ping";
    while (std::chrono::steady_clock::now() < deadline) {
        try {
            client.SendLatest(msg.data(), msg.size());
        } catch (const PeerClosed&) {
            threw = true;
            break;
        }
        SleepMs(10);
    }
    EXPECT_TRUE(threw);
}

TEST_F(SeqPacketRoundtripTest, SendDropsSilentlyWhenBufferFull) {
    SeqPacketEndpoint client = ConnectClient(tmp_.path(), 64, 1.0);
    SeqPacketEndpoint server_ep = server_->AcceptBlocking(1.0);

    // 送信バッファを極端に小さくして、受信側が読まない状態ですぐ満杯にする
    const int tiny = 1;
    ASSERT_EQ(::setsockopt(client.fd(), SOL_SOCKET, SO_SNDBUF, &tiny, sizeof(tiny)), 0);

    bool saw_drop = false;
    for (int i = 0; i < 50; ++i) {
        const std::string msg = "m" + std::to_string(i);
        if (!client.SendLatest(msg.data(), msg.size())) {
            saw_drop = true;
            break;
        }
    }
    EXPECT_TRUE(saw_drop) << "送信バッファ満杯時にfalseで破棄される設計だが、一度もfalseにならなかった";
}
